openpype/hosts/max/api/preview_animation.py

```diff
@@ -0,0 +1,310 @@
+        preview_res = rt.bitmap(
+            res_width, res_height, filename=filepath
+        )
+        dib = rt.gw.getViewportDib()
+        dib_width = float(dib.width)
+        dib_height = float(dib.height)
+        renderRatio = float(dib_width / dib_height)
+        if viewportRatio <= renderRatio:
+            heightCrop = (dib_width / renderRatio)
+            topEdge = int((dib_height - heightCrop) / 2.0)
+            tempImage_bmp = rt.bitmap(dib_width, heightCrop)
+            src_box_value = rt.Box2(0, topEdge, dib_width, heightCrop)
+        else:
+            widthCrop = dib_height * renderRatio
+            leftEdge = int((dib_width - widthCrop) / 2.0)
+            tempImage_bmp = rt.bitmap(widthCrop, dib_height)
+            src_box_value = rt.Box2(0, leftEdge, dib_width, dib_height)
+        rt.pasteBitmap(dib, tempImage_bmp, src_box_value, rt.Point2(0, 0))
+        # copy the bitmap and close it
+        rt.copy(tempImage_bmp, preview_res)
+        rt.close(tempImage_bmp)
+        rt.save(preview_res)
+        rt.close(preview_res)
+        rt.close(dib)
+        files.append(filepath)
+        if rt.keyboard.escPressed:
+            user_cancelled = True
+            break
+    # clean up the cache
+    rt.gc(delayed=True)
+    if user_cancelled:
+        raise RuntimeError("User cancelled rendering of viewport animation.")
+    return files
+
+
+def render_preview_animation(
+        filepath,
+        ext,
+        review_camera,
+        start_frame=None,
+        end_frame=None,
+        width=1920,
+        height=1080,
+        viewport_options=None):
+    """Render camera review animation
+    Args:
+        filepath (str): filepath to render to, without frame number and
+            extension
+        ext (str): output file extension
+        review_camera (str): viewport camera for preview render
+        start_frame (int): start frame
+        end_frame (int): end frame
+        width (int): render resolution width
+        height (int): render resolution height
+        viewport_options (dict): viewport setting options
+    Returns:
+        list: Rendered output files
+    """
+    if start_frame is None:
+        start_frame = int(rt.animationRange.start)
+    if end_frame is None:
+        end_frame = int(rt.animationRange.end)
+
+    if viewport_options is None:
+        viewport_options = viewport_options_for_preview_animation()
+    with play_preview_when_done(False):
+        with viewport_camera(review_camera):
+            with render_resolution(width, height):
+                if int(get_max_version()) < 2024:
+                    with viewport_preference_setting(
+                            viewport_options["general_viewport"],
+                            viewport_options["nitrous_viewport"],
+                            viewport_options["vp_btn_mgr"]
+                    ):
+                        percentSize = viewport_options.get("percentSize", 100)
```

Comment:
I'm a bit worried that this is **not** supported in the version for max 2024+ and thus the behavior will differ between max versions. Should we just always force percent size to 100 in the call. If the user would want to render a different resolution then they should just pass a different width x height. Yes?

Reply:
I agree with that to a certain extent, 2024Max has the precentSize parameter which can multiply the percentage to the resolution and it is deal withr ratio, but for the elder version, it is a bit different and it deosn't really deal with the ratio, we need to do some calculation to make it match the aspect ratio

Reply:
Sure - I just figured we either do not expose percent size and force the user of the function to put in whatever resolution they want. Otherwise if we really want to expose 'percentage' for the resolution we could maybe implement it as a fallback for the other case that does not support `percentSize` viewport options?

```python
percent = percentSize / 100.0
width = int(round(width * percent))
height = int(round(height * percent))
```

Reply:
> Sure - I just figured we either do not expose percent size and force the user of the function to put in whatever resolution they want. Otherwise if we really want to expose 'percentage' for the resolution we could maybe implement it as a fallback for the other case that does not support `percentSize` viewport options?
> 
> ```python
> percent = percentSize / 100.0
> width = int(round(width * percent))
> height = int(round(height * percent))
> ```

this is more easier to do the fallback with the elder version, Not sure about Max 2024+ as we used the maxscript for most of the viewport settings and `rt.Execute()` it.

Reply:
> this is more easier to do the fallback with the elder version, Not sure about Max 2024+ 

It's the older version that has the `percentSize` right - or are the function names reversed? It looks like `pre_max_2024` has the percent size usage.

Reply:
the new version also has the percentSize version, but the only difference is that being in the options list instead.

Reply:
See line 279

Reply:
I see 👍